custom_components/econnect_metronet/coordinator.py

```diff
@@ -0,0 +1,96 @@
+)
+from .devices import AlarmDevice
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AlarmCoordinator(DataUpdateCoordinator):
+    def __init__(self, hass: HomeAssistant, config: ConfigEntry) -> None:
+        # Initialize all clients and devices
+        client = ElmoClient(config.data[CONF_SYSTEM_URL], config.data[CONF_DOMAIN])
+        scan_interval = config.options.get(CONF_SCAN_INTERVAL, SCAN_INTERVAL_DEFAULT)
+
+        # Attributes
+        self.device = AlarmDevice(client, config.options)
+
+        # Configure the coordinator
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=scan_interval),
+        )
+
+    async def _async_update_data(self) -> Optional[Dict[str, Any]]:
+        """Update device data asynchronously using the long-polling method.
+
+        This method uses the e-Connect long-polling API implemented in `device.has_updates` which
+        blocks the thread for up to 15 seconds or until the backend pushes an update.
+        A timeout ensures the method doesn't hang indefinitely. In case of invalid token,
+        the method attempts to re-authenticate and get an update. If the update fails,
+        the connection is reset to ensure proper data alignment in subsequent runs.
+
+        Raises:
+            InvalidToken: When the token used for the connection is invalid.
+            UpdateFailed: When there's an error in updating the data.
+
+        Notes:
+            If the update fails, the method resets the device connection to ensure data
+            alignment and prevent the integration from getting stuck.
+        """
+        try:
+            if self.data is None:
+                # First update, no need to wait for changes
+                username = self.config_entry.data[CONF_USERNAME]
+                password = self.config_entry.data[CONF_PASSWORD]
+                await self.hass.async_add_executor_job(self.device.connect, username, password)
+                return await self.hass.async_add_executor_job(self.device.update)
+
+            async with async_timeout.timeout(POLLING_TIMEOUT):
+                # `device.has_updates` implements e-Connect long-polling API. This
+                # action blocks the thread for 15 seconds, or when the backend publishes an update
+                # POLLING_TIMEOUT ensures an upper bound regardless of the underlying implementation.
+                _LOGGER.debug("Coordinator | Waiting for changes (long-polling)")
+                status = await self.hass.async_add_executor_job(self.device.has_updates)
+                if status["has_changes"]:
+                    _LOGGER.debug("Coordinator | Changes detected, sending an update")
+                    return await self.hass.async_add_executor_job(self.device.update)
+                else:
+                    _LOGGER.debug("Coordinator | No changes detected")
+                    return {}
+        except InvalidToken:
```

Comment:
Add a comment to explain this is an exception that is expected at some point during the day. This is why we don't re-raise the exception (unless there is a further failure).